File: server/models/Player.js
const mongoose = require('mongoose');

const playerSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, 'Player name is required'],
      trim: true,
    },
    role: {
      type: String,
      required: [true, 'Player role is required'],
      enum: ['Batsman', 'Bowler', 'All-rounder', 'Wicketkeeper'],
    },
    basePrice: {
      type: Number,
      required: [true, 'Base price is required'],
      min: [0, 'Base price cannot be negative'],
    },
    sold: {
      type: Boolean,
      default: false,
    },
    teamId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Team',
      default: null,
    },
    price: {
      type: Number,
      default: null,
      min: [0, 'Price cannot be negative'],
    },
    nationality: {
      type: String,
      required: [true, 'Nationality is required'],
      trim: true,
    },
    age: {
      type: Number,
      required: [true, 'Age is required'],
      min: [16, 'Age must be at least 16'],
      max: [50, 'Age must be less than 50'],
    },
    battingStyle: {
      type: String,
      default: '',
    },
    bowlingStyle: {
      type: String,
      default: '',
    },
    image: {
      type: String,
      default: '',
    },
    stats: {
      matches: {
        type: Number,
        default: 0,
        min: 0,
      },
      runs: {
        type: Number,
        default: 0,
        min: 0,
      },
      wickets: {
        type: Number,
        default: 0,
        min: 0,
      },
      average: {
        type: Number,
        default: 0,
        min: 0,
      },
      strikeRate: {
        type: Number,
        default: 0,
        min: 0,
      },
    },
  },
  {
    timestamps: true,
  }
);

// Indexes for better query performance
playerSchema.index({ sold: 1 });
playerSchema.index({ role: 1 });
playerSchema.index({ teamId: 1 });
playerSchema.index({ name: 1 });

// Virtual for generating default avatar
playerSchema.virtual('avatarUrl').get(function () {
  if (this.image) {
    return this.image;
  }
  return `https://ui-avatars.com/api/?name=${encodeURIComponent(this.name)}&background=6366f1&color=fff&size=128`;
});

// Ensure virtuals are included when converting to JSON
playerSchema.set('toJSON', { virtuals: true });
playerSchema.set('toObject', { virtuals: true });

const Player = mongoose.model('Player', playerSchema);

module.exports = Player;

File: backend/app/schemas/player.py
from pydantic import BaseModel
from typing import Optional


class PlayerStats(BaseModel):
    matches: Optional[int] = 0
    runs: Optional[int] = 0
    wickets: Optional[int] = 0
    average: Optional[float] = 0.0
    strikeRate: Optional[float] = 0.0


class PlayerBase(BaseModel):
    name: str
    role: str
    basePrice: int
    nationality: str
    age: int
    battingStyle: Optional[str] = None
    bowlingStyle: Optional[str] = None
    image: str
    stats: Optional[PlayerStats] = None


class PlayerCreate(PlayerBase):
    pass


class PlayerUpdate(BaseModel):
    name: Optional[str] = None
    role: Optional[str] = None
    basePrice: Optional[int] = None
    nationality: Optional[str] = None
    age: Optional[int] = None
    battingStyle: Optional[str] = None
    bowlingStyle: Optional[str] = None
    image: Optional[str] = None
    stats: Optional[PlayerStats] = None


class Player(PlayerBase):
    id: int
    sold: bool = False
    teamId: Optional[int] = None
    price: Optional[int] = None

    class Config:
        from_attributes = True
